protocol/bc/hash.go

```diff
@@ -1,103 +1,92 @@
 package bc
 
 import (
-	"bytes"
 	"database/sql/driver"
-	"encoding/hex"
-	"encoding/json"
-	"fmt"
+	"encoding/binary"
 	"io"
 
 	"golang.org/x/crypto/sha3"
 
 	"chain/crypto/sha3pool"
 	"chain/encoding/blockchain"
-	"chain/errors"
 )
 
-// Hash represents a 256-bit hash.  By convention, Hash objects are
-// typically passed as values, not as pointers.
-type Hash [32]byte
+var EmptyStringHash Hash
 
-var EmptyStringHash = sha3.Sum256(nil)
+func init() {
+	EmptyStringHash.FromByte32(sha3.Sum256(nil))
+}
+
+// Hash represents a 256-bit hash.  Data structure defined in
+// hash.proto.
 
-// String returns the bytes of h encoded in hex.
-func (h Hash) String() string {
-	b, _ := h.MarshalText() // #nosec
-	return string(b)
+func (h Hash) Byte32() (b32 Byte32) {
```

Comment:
And likewise, could this be

```go
func (h Hash) Byte32() [32]byte
```

(Is this even actually used anywhere? I don't see it.)